File: backend/internal/application/billing/schedule_pricing.go
package billing

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/DEEIX-AI/DEEIX-Chat/backend/internal/repository"
)

// 时段倍率（峰谷计费）：在模型基础单价之上按服务器本地时间叠加倍率百分比，与计费模式正交，
// token / call / duration / tiered 都适用。同一模型的时段不允许重叠，因此任一时刻最多命中一条。

const (
	maxSchedulePeriods       = 32
	maxSchedulePeriodName    = 32
	minScheduleRatePercent   = 1
	maxScheduleRatePercent   = 10000
	minutesPerDay            = 24 * 60
	schedulePeriodTimeLayout = "15:04"
)

type schedulePricingConfig struct {
	Periods []schedulePeriod `json:"periods"`
}

// schedulePeriod 是一条时段规则。Start/End 为 "HH:MM"，End 不大于 Start 表示跨过午夜；
// Weekdays 为 0（周日）到 6（周六），按 Start 所在的自然日判断。
type schedulePeriod struct {
	Name        string `json:"name"`
	Weekdays    []int  `json:"weekdays"`
	Start       string `json:"start"`
	End         string `json:"end"`
	RatePercent int    `json:"ratePercent"`

	startMinute int
	endMinute   int
}

// ResolvedSchedulePeriod 是某个时刻命中的时段，供账本快照与用户展示。
type ResolvedSchedulePeriod struct {
	Name        string
	RatePercent int
}

func parseClockMinute(value string) (int, error) {
	parsed, err := time.Parse(schedulePeriodTimeLayout, strings.TrimSpace(value))
	if err != nil {
		return 0, repository.ErrInvalidInput
	}
	return parsed.Hour()*60 + parsed.Minute(), nil
}

// parseSchedulePeriods 解析并校验时段配置；空配置表示未启用。
func parseSchedulePeriods(raw string) ([]schedulePeriod, error) {
	raw = strings.TrimSpace(raw)
	if raw == "" || raw == "{}" {
		return nil, nil
	}
	var config schedulePricingConfig
	if err := json.Unmarshal([]byte(raw), &config); err != nil {
		return nil, repository.ErrInvalidInput
	}
	if len(config.Periods) == 0 {
		return nil, nil
	}
	if len(config.Periods) > maxSchedulePeriods {
		return nil, repository.ErrInvalidInput
	}
	for index := range config.Periods {
		period := &config.Periods[index]
		period.Name = strings.TrimSpace(period.Name)
		if period.Name == "" || len([]rune(period.Name)) > maxSchedulePeriodName {
			return nil, repository.ErrInvalidInput
		}
		if period.RatePercent < minScheduleRatePercent || period.RatePercent > maxScheduleRatePercent {
			return nil, repository.ErrInvalidInput
		}
		if len(period.Weekdays) == 0 {
			return nil, repository.ErrInvalidInput
		}
		seen := make(map[int]struct{}, len(period.Weekdays))
		for _, weekday := range period.Weekdays {
			if weekday < 0 || weekday > 6 {
				return nil, repository.ErrInvalidInput
			}
			if _, duplicate := seen[weekday]; duplicate {
				return nil, repository.ErrInvalidInput
			}
			seen[weekday] = struct{}{}
		}
		sort.Ints(period.Weekdays)
		start, err := parseClockMinute(period.Start)
		if err != nil {
			return nil, err
		}
		end, err := parseClockMinute(period.End)
		if err != nil {
			return nil, err
		}
		if start == end {
			return nil, repository.ErrInvalidInput
		}
		period.Start = fmt.Sprintf("%02d:%02d", start/60, start%60)
		period.End = fmt.Sprintf("%02d:%02d", end/60, end%60)
		period.startMinute = start
		period.endMinute = end
	}
	if schedulePeriodsOverlap(config.Periods) {
		return nil, repository.ErrInvalidInput
	}
	return config.Periods, nil
}

// weekMinuteSpans 把一条时段展开为一周内的分钟区间 [from, to)，跨午夜的拆成两段，周日晚跨到周一。
func weekMinuteSpans(period schedulePeriod) [][2]int {
	spans := make([][2]int, 0, len(period.Weekdays)*2)
	for _, weekday := range period.Weekdays {
		base := weekday * minutesPerDay
		if period.startMinute < period.endMinute {
			spans = append(spans, [2]int{base + period.startMinute, base + period.endMinute})
			continue
		}
		spans = append(spans, [2]int{base + period.startMinute, base + minutesPerDay})
		next := ((weekday + 1) % 7) * minutesPerDay
		spans = append(spans, [2]int{next, next + period.endMinute})
	}
	return spans
}

func schedulePeriodsOverlap(periods []schedulePeriod) bool {
	type span struct {
		from, to int
	}
	all := make([]span, 0, len(periods)*4)
	for _, period := range periods {
		for _, item := range weekMinuteSpans(period) {
			if item[0] < item[1] {
				all = append(all, span{from: item[0], to: item[1]})
			}
		}
	}
	sort.Slice(all, func(i, j int) bool { return all[i].from < all[j].from })
	for index := 1; index < len(all); index++ {
		if all[index].from < all[index-1].to {
			return true
		}
	}
	return false
}

// normalizeSchedulePricingJSON 空配置落库为 "{}"，与列默认值一致。
func normalizeSchedulePricingJSON(raw string) (string, error) {
	periods, err := parseSchedulePeriods(raw)
	if err != nil {
		return "", err
	}
	if len(periods) == 0 {
		return "{}", nil
	}
	payload, err := json.Marshal(schedulePricingConfig{Periods: periods})
	if err != nil {
		return "", err
	}
	return string(payload), nil
}

// resolveSchedulePeriod 返回 at（按服务器本地时间）命中的时段；未命中返回 nil。
func resolveSchedulePeriod(periods []schedulePeriod, at time.Time) *ResolvedSchedulePeriod {
	if len(periods) == 0 {
		return nil
	}
	local := at.Local()
	weekMinute := int(local.Weekday())*minutesPerDay + local.Hour()*60 + local.Minute()
	for _, period := range periods {
		for _, span := range weekMinuteSpans(period) {
			if weekMinute >= span[0] && weekMinute < span[1] {
				return &ResolvedSchedulePeriod{Name: period.Name, RatePercent: period.RatePercent}
			}
		}
	}
	return nil
}

// resolveSchedulePeriodJSON 是账本与预估共用的入口：解析失败时视为无时段，不阻断计费。
func resolveSchedulePeriodJSON(raw string, at time.Time) *ResolvedSchedulePeriod {
	periods, err := parseSchedulePeriods(raw)
	if err != nil {
		return nil
	}
	return resolveSchedulePeriod(periods, at)
}

File: frontend/shared/model/schedule-pricing.ts
import { formatRateMultiplier as formatMultiplier } from "@/shared/lib/billing-display";

// Time-of-day rate multipliers ("峰谷") attached to a model pricing entry.
// Mirrors backend/internal/application/billing/schedule_pricing.go: periods use
// the server's local clock, may wrap past midnight, and must not overlap.

export type SchedulePeriodForm = {
  id: string;
  name: string;
  weekdays: number[];
  start: string;
  end: string;
  ratePercent: string;
};

export type SchedulePeriodPayload = {
  name: string;
  weekdays: number[];
  start: string;
  end: string;
  ratePercent: number;
};

export const SCHEDULE_PERIOD_LIMITS = { maxPeriods: 32, nameLength: 32, minRatePercent: 1, maxRatePercent: 10000 } as const;
export const WEEKDAY_ORDER = [1, 2, 3, 4, 5, 6, 0] as const;
export const ALL_WEEKDAYS = [0, 1, 2, 3, 4, 5, 6] as const;

const MINUTES_PER_DAY = 24 * 60;
const TIME_RE = /^([01]?\d|2[0-3]):([0-5]\d)$/;

let periodSequence = 0;
export function createSchedulePeriodForm(partial?: Partial<Omit<SchedulePeriodForm, "id">>): SchedulePeriodForm {
  periodSequence += 1;
  return {
    id: `period-${Date.now()}-${periodSequence}`,
    name: "",
    weekdays: [1, 2, 3, 4, 5],
    start: "09:00",
    end: "18:00",
    ratePercent: "100",
    ...partial,
  };
}

export function parseClockMinute(value: string): number | null {
  const match = TIME_RE.exec(value.trim());
  if (!match) {
    return null;
  }
  return Number(match[1]) * 60 + Number(match[2]);
}

// Tolerant reader for hand-typed times, used to canonicalise an input on blur:
// "830" → 08:30, "8:00" → 08:00, "13:5" → 13:05. Null when unreadable.
export function parseClockLoose(value: string): number | null {
  // CJK input can arrive as full-width digits and punctuation.
  const text = value
    .trim()
    .replace(/[０-９]/g, (digit) => String.fromCharCode(digit.charCodeAt(0) - 0xfee0))
    .replace(/[：.、]/g, ":")
    .replace(/\s+/g, "");
  if (text === "") {
    return null;
  }
  const [, hourText = "", minuteText = ""] = /^(\d{1,2}):(\d{1,2})$/.exec(text) ?? [];
  if (hourText) {
    const hour = Number(hourText);
    const minute = Number(minuteText);
    if (minute > 59) {
      return null;
    }
    // 24:00 is the usual way to write "until midnight"; store it as 00:00.
    if (hour === 24 && minute === 0) {
      return 0;
    }
    return hour <= 23 ? hour * 60 + minute : null;
  }
  if (!/^\d{1,4}$/.test(text)) {
    return null;
  }
  // Digits only: 1–2 digits are an hour, 3 digits are H:MM, 4 digits are HH:MM.
  const padded = text.padStart(4, "0");
  const hour = Number(text.length <= 2 ? text : padded.slice(0, 2));
  const minute = text.length <= 2 ? 0 : Number(padded.slice(2));
  return hour <= 23 && minute <= 59 ? hour * 60 + minute : null;
}

// Canonical HH:MM for an input field; unreadable text is kept as typed so the
// save-time validator can still complain about it.
export function canonicalClock(value: string): string {
  const minute = parseClockLoose(value);
  return minute === null ? value.trim() : formatClock(minute);
}

// An end not after the start means the period runs past midnight.
export function isOvernightPeriod(period: Pick<SchedulePeriodPayload, "start" | "end">): boolean {
  const start = parseClockLoose(period.start);
  const end = parseClockLoose(period.end);
  return start !== null && end !== null && end <= start;
}

function formatClock(minute: number): string {
  return `${String(Math.floor(minute / 60)).padStart(2, "0")}:${String(minute % 60).padStart(2, "0")}`;
}

function isSchedulePeriodPayload(value: unknown): value is SchedulePeriodPayload {
  if (!value || typeof value !== "object") {
    return false;
  }
  const record = value as Record<string, unknown>;
  return (
    typeof record.name === "string" &&
    Array.isArray(record.weekdays) &&
    record.weekdays.every((day) => Number.isInteger(day) && (day as number) >= 0 && (day as number) <= 6) &&
    typeof record.start === "string" &&
    typeof record.end === "string" &&
    typeof record.ratePercent === "number"
  );
}

// Accepts the stored JSON string or an already-parsed object; null when absent or malformed.
export function parseSchedulePricing(raw: unknown): SchedulePeriodPayload[] | null {
  try {
    const parsed: unknown = typeof raw === "string" ? (raw.trim() ? JSON.parse(raw) : null) : raw;
    if (!parsed || typeof parsed !== "object" || !("periods" in parsed) || !Array.isArray(parsed.periods)) {
      return null;
    }
    const periods = parsed.periods.filter(isSchedulePeriodPayload);
    return periods.length === parsed.periods.length ? periods : null;
  } catch {
    return null;
  }
}

export function schedulePeriodsToForm(periods: SchedulePeriodPayload[] | null | undefined): SchedulePeriodForm[] {
  return (periods ?? []).map((period) =>
    createSchedulePeriodForm({
      name: period.name,
      weekdays: [...period.weekdays],
      start: period.start,
      end: period.end,
      ratePercent: String(period.ratePercent),
    }),
  );
}

// Week-minute spans [from, to) for overlap checks; a wrap-around period yields two spans.
function weekSpans(period: SchedulePeriodPayload): [number, number][] {
  const start = parseClockMinute(period.start) ?? 0;
  const end = parseClockMinute(period.end) ?? 0;
  const spans: [number, number][] = [];
  for (const weekday of period.weekdays) {
    const base = weekday * MINUTES_PER_DAY;
    if (start < end) {
      spans.push([base + start, base + end]);
      continue;
    }
    spans.push([base + start, base + MINUTES_PER_DAY]);
    const next = ((weekday + 1) % 7) * MINUTES_PER_DAY;
    if (end > 0) {
      spans.push([next, next + end]);
    }
  }
  return spans;
}

export type SchedulePeriodIssue = "name" | "weekdays" | "time" | "rate" | "overlap" | "count";

// Validates and normalises the editor rows into payloads; issues are keyed by row id ("" for list-level).
export function normalizeSchedulePeriods(rows: SchedulePeriodForm[]): { periods: SchedulePeriodPayload[]; issues: Map<string, SchedulePeriodIssue> } {
  const issues = new Map<string, SchedulePeriodIssue>();
  const periods: SchedulePeriodPayload[] = [];
  if (rows.length > SCHEDULE_PERIOD_LIMITS.maxPeriods) {
    issues.set("", "count");
  }
  for (const row of rows) {
    const name = row.name.trim();
    const start = parseClockMinute(row.start);
    const end = parseClockMinute(row.end);
    const ratePercent = Number(row.ratePercent);
    if (!name || name.length > SCHEDULE_PERIOD_LIMITS.nameLength) {
      issues.set(row.id, "name");
    } else if (row.weekdays.length === 0) {
      issues.set(row.id, "weekdays");
    } else if (start === null || end === null || start === end) {
      issues.set(row.id, "time");
    } else if (!Number.isInteger(ratePercent) || ratePercent < SCHEDULE_PERIOD_LIMITS.minRatePercent || ratePercent > SCHEDULE_PERIOD_LIMITS.maxRatePercent) {
      issues.set(row.id, "rate");
    } else {
      periods.push({ name, weekdays: [...row.weekdays].sort((a, b) => a - b), start: formatClock(start), end: formatClock(end), ratePercent });
    }
  }
  // periods[i] came from validRows[i]; resolve before marking overlaps so indices stay aligned.
  const validRows = rows.filter((item) => !issues.has(item.id));
  const spans = periods.flatMap((period, index) => weekSpans(period).map((span) => ({ span, index })));
  spans.sort((a, b) => a.span[0] - b.span[0]);
  for (let index = 1; index < spans.length; index += 1) {
    const previous = spans[index - 1];
    const current = spans[index];
    if (previous && current && current.span[0] < previous.span[1]) {
      const row = validRows[current.index];
      if (row) {
        issues.set(row.id, "overlap");
      }
    }
  }
  return { periods, issues };
}

export function stringifySchedulePricing(periods: SchedulePeriodPayload[]): string {
  return periods.length === 0 ? "" : JSON.stringify({ periods });
}

// Which period applies right now on the server clock, expressed via its UTC offset.
export function resolveCurrentSchedulePeriod(periods: readonly SchedulePeriodPayload[], serverUTCOffsetMinutes: number, now = new Date()): SchedulePeriodPayload | null {
  if (periods.length === 0) {
    return null;
  }
  const serverNow = new Date(now.getTime() + (serverUTCOffsetMinutes + now.getTimezoneOffset()) * 60_000);
  const weekMinute = serverNow.getDay() * MINUTES_PER_DAY + serverNow.getHours() * 60 + serverNow.getMinutes();
  for (const period of periods) {
    for (const [from, to] of weekSpans(period)) {
      if (weekMinute >= from && weekMinute < to) {
        return period;
      }
    }
  }
  return null;
}

export function formatRateMultiplier(ratePercent: number): string {
  return formatMultiplier(ratePercent / 100);
}
